### PR TITLE
8256308: Send arguments to javac server in a config file

### PR DESCRIPTION
**Pull request by magicus**
Currently, to use the javac server, a horrendously long command line option is created, looking like this: `--server:portfile=<path to portfile>:sjavac=<command to launch server>`, where the sjavac command has had all spaces replaced by %20. Since Project Jigsaw, the set of module arguments needed is huge to begin with, making this command line incomprehensible after mangling.

Apart from making java command lines hard to read (and copy/paste!) by developers, it also makes it hard for scripts to parse. The upcoming winenv rewrite is dependent on being able to differentiate between path names and other arguments, which is not possible in this mess.

So, instead, let's write it to a file, without any escaping, and just pass the configuration file name to the server.

Note that this will change the behavior of the javac server, but as the source code states this is not a documented or externally supported API no CSR is needed. 

I also cleaned up some code in SjavacClient, in particular code relating to the passing of arguments. (We never change poolsize or keepalive when we call it.)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256308](https://bugs.openjdk.java.net/browse/JDK-8256308): Send arguments to javac server in a config file

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**)
 * [Joel Borggrén-Franck](https://openjdk.java.net/census#jfranck) (@jbf - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1195/head:pull/1195`
`$ git checkout pull/1195`

### CODE REVIEW

File: make/common/JavaCompilation.gmk [hunk 1]
```diff
@@ -215,18 +221,35 @@ define SetupJavaCompilationBody
...
       # and the cookie necessary to talk to the server.
-      $1_JAVA_SERVER_FLAGS := --server:portfile=$$(JAVAC_SERVER_DIR)/server.port,sjavac=$$($1_ESCAPED_CMD)
+      $1_JAVAC_PORT_FILE := $$(JAVAC_SERVER_DIR)/server.port
+
+      # The servercmd specified how to launch the server. This must use
+      # JAVA_DETACH, which is the "big" java with an ability to detach from
+      # fixpath (on Windows) This will be executed by the client, if needed.
+      $1_JAVAC_SERVER_CMD := $$(JAVA_DETACH) $$($1_JAVA_FLAGS) $$($1_JAVAC)
+
+      $1_CONFIG_VARDEPS :=$$($1_JAVAC_PORT_FILE) $$($1_JAVAC_SERVER_CMD)
```
Comment (erikj79) [inline make/common/JavaCompilation.gmk:237]: Space

File: make/common/JavaCompilation.gmk [hunk 2]
```diff
@@ -215,18 +221,35 @@ define SetupJavaCompilationBody
...
+      $1_CONFIG_VARDEPS_FILE := $$(call DependOnVariable, $1_CONFIG_VARDEPS, \
+        $$($1_BIN)$$($1_MODULE_SUBDIR)/_the.$1.config_vardeps)
```
Comment (erikj79) [inline make/common/JavaCompilation.gmk:239]: Indentation 4

File: make/common/JavaCompilation.gmk [hunk 3]
```diff
@@ -215,18 +221,35 @@ define SetupJavaCompilationBody
...
 
-      # Always use small to launch client
-      $1_JAVAC_CMD := $$(JAVA_SMALL) $$($1_JAVA_FLAGS) $$($1_JAVAC) $$($1_JAVA_SERVER_FLAGS)
+      ifeq ($(call isBuildOs, windows), true)
+        $1_ECHO_COMMAND := $(FIXPATH) cmd /c echo
+      else
+        $1_ECHO_COMMAND := $(ECHO)
+      endif
+              $$($1_JAVAC_SERVER_CONFIG): $$($1_CONFIG_VARDEPS_FILE)
```
Comment (erikj79) [inline make/common/JavaCompilation.gmk:246]: Indentation?

**Comment by erikj79** [comment]
Looks good in general, just some whitespace issues.

**Review by erikj79 — APPROVED** [review]

**Review by jonathan-gibbons — COMMENTED** [review]
There seems to be much more here than there needs to be.

If the issue is just long command lines, then the obvious/conventional solution would be to use @files, which would be a tiny change to the sjavac source code, to insert a single call to `CommandLine.parse` to expand any @file arguments on the command line.

So, before reading all the various details in this proposed change, is there any reason why the simple @file solution cannot be used?

**Comment by magicus** [comment]
The main issue is not the long command lines. Getting shorter command lines is nice, but more a side-effect.

The real driver here is the awkward encoding of the command line, with spaces replaced by %20. This makes it impossible to parse the --server option to Java and properly replace unix paths with Windows paths, in the way that is needed by the upcoming changes to how we build on Windows.

File: src/jdk.compiler/share/classes/com/sun/tools/sjavac/Util.java [hunk 4]
```diff
@@ -79,15 +79,23 @@ public static String extractStringOption(String opName, String s) {
...
         if (pe == -1) pe = s.length();
         return s.substring(p, pe);
     }
 
+    public static String extractStringOption(String opName, String s, String deflt) {
+        return extractStringOptionWithDelimiter(opName, s, deflt, ',');
+    }
+
+    public static String extractStringOptionLine(String opName, String s, String deflt) {
+        return extractStringOptionWithDelimiter(opName, s, deflt, '\n').strip();
```
Comment (jbf) [inline src/jdk.compiler/share/classes/com/sun/tools/sjavac/Util.java:96]: Is '\n' going to be problematic due to differences in line endings on various platforms?

**Review by jbf — APPROVED** [review]

File: src/jdk.compiler/share/classes/com/sun/tools/sjavac/Util.java (hunk 4 above)
Comment (magicus) [reply src/jdk.compiler/share/classes/com/sun/tools/sjavac/Util.java:96]: The JDK project is limited to unix-style line endings only. No other styles are supported, and any attempt to use such will break the build (often early on).

**Pull request closed without merging**